src/frontend/src/optimizer/plan_node/stream_materialize.rs

```diff
@@ -99,17 +99,25 @@ impl StreamMaterialize {
     }
 
     /// Create a materialize node.
+    ///
+    /// When creating index, `distribute_only_order_by` should be true. We should distribute keys
+    /// using order by columns, instead of pk.
     pub fn create(
         input: PlanRef,
         mv_name: String,
         user_order_by: Order,
         user_cols: FixedBitSet,
+        distribute_only_order_by: bool,
     ) -> Result<Self> {
         // ensure the same pk will not shuffle to different node
         let input = match input.distribution() {
             Distribution::Single => input,
-            _ => Distribution::HashShard(input.pk_indices().to_vec())
-                .enforce_if_not_satisfies(input, Order::any()),
+            _ => Distribution::HashShard(if distribute_only_order_by {
```

Comment:
Looks somehow strange to me... I think this should be handled by the caller, not here.

Reply:
Infer distribution is always done inside operators?

Reply:
Hmmm. Also makes sense

Reply:
In fact I'm also considering making `Materialize` to be `AnyShard`, and let the create mv / index statement to decide the distribution by specifying it in PlanRoot. But in this case, we might need some redesign.